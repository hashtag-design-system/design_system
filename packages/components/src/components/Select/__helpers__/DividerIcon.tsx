import React from "react";

export const DividerIcon: React.FC = () => {
  return (
    <svg width={2} className="select__input__divider" height={34} fill="none">
      <path d="M1 32V1" strokeWidth={1.5} strokeLinecap="round" />
    </svg>
  );
};
