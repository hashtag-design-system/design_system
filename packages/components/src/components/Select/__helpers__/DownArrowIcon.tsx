import { motion } from "framer-motion";
import React, { useContext } from "react";
import DropdownContext from "../../../utils/ctx/DropdownContext";

export const DownArrowIcon: React.FunctionComponent = () => {
  const selectCtx = useContext(DropdownContext);

  const { isVisible, handleClick } = selectCtx;

  // https://blog.lftechnology.com/using-svg-icons-components-in-react-44fbe8e5f91
  return (
    <motion.svg
      fill="none"
      viewBox="0 0 24 24"
      className="input__icon select__icon"
      // * A "spring" animation might be too much and useless for this element
      animate={isVisible ? { rotate: 180 } : { rotate: 0 }}
      onClick={e => handleClick && handleClick(e)}
    >
      <path d="M21 7l-9 10L3 7" strokeWidth={1.5} strokeLinecap="round" strokeLinejoin="round" />
    </motion.svg>
  );
};
