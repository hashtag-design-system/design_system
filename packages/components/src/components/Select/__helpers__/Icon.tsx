import { motion, SVGMotionProps } from "framer-motion";
import React from "react";

export const Icon: React.FC<{ clicked: boolean; onClick?: SVGMotionProps<SVGElement>["onClick"] }> = ({ clicked, ...props }) => {
  return (
    <motion.svg
      width={24}
      height={24}
      fill="none"
      className="input__icon select__icon"
      animate={clicked ? { rotate: 180 } : { rotate: 0 }}
      // * A "spring" animation might be too much and useless for this element
      {...props}
    >
      <path d="M21 7l-9 10L3 7" stroke="var(--grey-7)" strokeWidth={2} strokeLinecap="round" strokeLinejoin="round" />
    </motion.svg>
  );
};

export default Icon;
