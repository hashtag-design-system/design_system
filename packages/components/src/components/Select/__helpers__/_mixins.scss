@mixin select__item--hover() {
  outline: none;

  &:not(.select__header) {
    background-color: $grey-4;
    cursor: pointer;
  }
}

.select__item--hover {
  @include select__item--hover();
}
