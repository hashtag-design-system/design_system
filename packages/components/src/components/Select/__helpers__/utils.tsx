import { render, RenderOptions } from "@testing-library/react";
import React from "react";
import { SelectContextProvider, SelectContextType } from "../../../utils";

export const defaultProps: Pick<
  SelectContextType,
  "isOpen" | "isMobile" | "multiSelectable" | "items" | "value" | "width" | "filterValue"
> = {
  isOpen: false,
  isMobile: false,
  multiSelectable: false,
  value: "",
  items: [],
  width: "200px",
  filterValue: [],
};

type SelectCustomRenderOptions = {
  providerProps?: Partial<SelectContextType>;
  renderOptions?: RenderOptions;
};

export const selectCustomRender = (ui: React.ReactElement, options?: SelectCustomRenderOptions) => {
  return render(
    // @ts-expect-error
    <SelectContextProvider value={{ ...defaultProps, ...options?.providerProps }}>{ui}</SelectContextProvider>,
    options?.renderOptions
  );
};
