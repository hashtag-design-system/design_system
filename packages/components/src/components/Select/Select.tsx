import React, { useEffect, useRef, useState } from "react";
import { SelectContextProvider } from "../../utils/contexts";
import { useClassnames, useClickOutside, useIsMobile } from "../../utils/hooks";
import { ComponentProps } from "../__helpers__";
import { Button } from "./Button";
import Countries from "./Countries";
import { Filter } from "./Filter";
import { Header } from "./Header";
import { Hr } from "./Hr";
import { Item } from "./Item";
import { Modal } from "./Modal";
import { Options } from "./Options";

export type SelectedItems = {
  id: string;
  content: string;
  valueAlternative?: string;
  highlightedChildren: string;
  selected: boolean;
  isShown: boolean;
  ref?: React.RefObject<HTMLDivElement>;
};

export type Props = {
  defaultOpen?: boolean;
  multiSelectable?: boolean;
  mobileView?: boolean;
  onSelect?: (selected: SelectedItems[]) => void;
  onDismiss?: (e: React.MouseEvent<HTMLElement>) => void;
  onOutsideClick?: (outsideClick: boolean) => void;
} & Pick<React.CSSProperties, "width">;

type SubComponents = {
  Item: typeof Item;
  Header: typeof Header;
  Modal: typeof Modal;
  Button: typeof Button;
  Hr: typeof Hr;
  Filter: typeof Filter;
  Options: typeof Options;
  Countries: typeof Countries;
};

export type FProps = Props &
  Omit<ComponentProps<"details", false>, "onSelect"> & {
    forwardRef?: ComponentProps<"details", true>["ref"];
  };

export type SBProps = Props & Pick<FProps, "defaultOpen" | "onToggle" | "onSelect">;

const Select: React.FC<FProps> & SubComponents = ({
  defaultOpen = false,
  multiSelectable = false,
  mobileView = false,
  children,
  open,
  forwardRef,
  width = "200px",
  onToggle,
  onSelect,
  onOutsideClick,
  onDismiss,
  ...props
}) => {
  const { ref: modalRef, isOpen, setIsOpen, outsideClick } = useClickOutside<HTMLDivElement>(defaultOpen, undefined, onDismiss);
  const [value, setValue] = useState<string>("");
  const [items, setItems] = useState<SelectedItems[]>([]);
  const [isDisabled, setIsDisabled] = useState(false);
  const [classNames, rest] = useClassnames("select__box__container", props);
  const { isMobile } = useIsMobile(mobileView);

  useEffect(() => {
    if (onOutsideClick) {
      onOutsideClick(outsideClick);
    }
  }, [outsideClick, onOutsideClick]);

  const divRef = useRef<HTMLDivElement>(null);

  const handleClick = (e: React.MouseEvent<HTMLElement>) => {
    if (isDisabled) {
      e.preventDefault();
    }
  };

  const handleToggle = (e: React.SyntheticEvent<HTMLElement>, boolean = true) => {
    e.preventDefault();

    if (!isDisabled) {
      if (boolean) {
        // @ts-expect-error
        setIsOpen(e.target.open);
      } else {
        setIsOpen(!isOpen);
      }
    }

    if (onToggle) {
      onToggle(e);
    }
  };

  const handleKeyDown = (e: React.KeyboardEvent<HTMLElement>) => {
    if (e.target instanceof HTMLInputElement) {
      return;
    }
    e.preventDefault();

    const focusItem = (next: boolean) => {
      if (divRef && divRef.current) {
        const options = Array.from(divRef.current.getElementsByClassName("select__item"));
        const selected = document.activeElement;

        let item: Element = options[0];
        if (selected && selected.tagName.toLowerCase() === "div") {
          const idx = options.indexOf(selected);
          if (next) {
            const nextOpt = options[idx + 1];
            if (nextOpt) {
              item = nextOpt;
              // (nextOpt as HTMLDivElement).focus();
            }
          } else {
            const nextOpt = options[idx - 1];
            if (nextOpt) {
              item = nextOpt;
              // (nextOpt as HTMLDivElement).focus();
            } else {
              item = options[options.length - 1];
            }
          }
        }
        return item;
      }
    };

    const isSummaryFocused = e.target instanceof Element && e.target.tagName.toLowerCase() === "summary";
    switch (e.code) {
      case "Escape":
        if (isOpen) {
          handleToggle(e, false);
        }
        break;
      case "Space":
      case "Enter":
        handleToggle(e, false);

        const selected = document.activeElement;
        if (!isSummaryFocused) {
          if (selected && selected.className === "select__item") {
            // @ts-expect-error
            document.activeElement.click();
            e.preventDefault();
          }
        }

        break;
      case "ArrowDown":
        if (isSummaryFocused && !isOpen) {
          handleToggle(e, false);
        } else {
          const target = focusItem(true);
          if (target) {
            (target as HTMLDivElement).focus();
          }
          e.preventDefault();
        }
        break;
      case "ArrowUp":
        if (isSummaryFocused && isOpen) {
          handleToggle(e, false);
        } else {
          const target = focusItem(false);
          if (target) {
            (target as HTMLDivElement).focus();
          }
          e.preventDefault();
        }
        break;
    }
  };

  useEffect(() => {
    setValue(
      items
        .filter(item => item.selected)
        .map(item => {
          if (item.valueAlternative) {
            return item.valueAlternative;
          } else {
            return item.content;
          }
        })
        .join(", ")
    );
  }, [items]);

  const fOpen = open === undefined ? isOpen : open;
  const fMobile = isMobile || mobileView;

  return (
    <SelectContextProvider
      value={{
        isOpen: fOpen,
        ref: forwardRef,
        value,
        multiSelectable,
        items,
        isMobile: fMobile,
        modalRef,
        width,
        setItems,
        handleToggle,
        setIsDisabled,
        onSelect,
      }}
    >
      <div className="select__container" ref={divRef} data-testid="select-container">
        <details
          ref={fMobile ? undefined : modalRef}
          className={classNames}
          open={fOpen}
          onClick={e => handleClick(e)}
          onToggle={e => handleToggle(e)}
          onKeyDown={e => handleKeyDown(e)}
          data-testid="select"
          {...rest}
        >
          {children}
        </details>
      </div>
    </SelectContextProvider>
  );
};

Select.Header = Header;
Select.Item = Item;
Select.Modal = Modal;
Select.Button = Button;
Select.Hr = Hr;
Select.Filter = Filter;
Select.Options = Options;
Select.Countries = Countries;

export default Select;
