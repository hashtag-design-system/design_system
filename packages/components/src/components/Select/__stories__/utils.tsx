import Select from "../index";

export type StoriesComponentProps = { headerChildren?: boolean; optionsChilden?: boolean };

export const StoriesItems: React.FC<Pick<StoriesComponentProps, "optionsChilden">> = ({ optionsChilden = false }) => {
  return (
    <>
      {!optionsChilden && <Select.Filter placeholder="Filter" floatingplaceholder={false} />}
      <Select.Options>
        {optionsChilden && <Select.Filter placeholder="Filter" floatingplaceholder={false} />}
        <Select.Item id="hey_george" content="Hey george" />
        <Select.Item id="amsterdam" content="Amsterdam" htmlContent={{ after: <p>, NL</p> }} />
        <Select.Item id="georgekrax" content="georgekrax" />
        <Select.Item id="georgekrax2" content="georgekrachropoulos" />
        <Select.Item id="test" content="Test item" />
        <Select.Item id="test_id" content=" Amsterdam" htmlContent={{ before: <strong>NL</strong> }} />
      </Select.Options>
    </>
  );
};

export const StoriesComponent: React.FC<any & StoriesComponentProps> = ({
  headerChildren = false,
  optionsChilden = false,
  align = "left",
  ...props
}) => {
  return (
    <Select {...props}>
      <Select.Button style={{ width: "200px" }}>Projects</Select.Button>
      <Select.Modal align={align}>
        {headerChildren ? (
          <Select.Header value="header" {...props}>
            <StoriesItems optionsChilden={optionsChilden} />
          </Select.Header>
        ) : (
          <>
            <Select.Header value="header" {...props} />
            <StoriesItems optionsChilden={optionsChilden} />
          </>
        )}
      </Select.Modal>
    </Select>
  );
};
