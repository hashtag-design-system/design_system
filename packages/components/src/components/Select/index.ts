import Select from "./Select";

export type { FProps as SelectButtonFProps, Props as SelectButtonProps, SBProps as SelectButtonSBProps } from "./Button";
export type { FProps as SelectFilterFProps, Props as SelectFilterProps } from "./Filter";
export type { FProps as SelectHeaderFProps, Props as SelectHeaderProps } from "./Header";
export type { FProps as SelectHrFProps } from "./Hr";
export type { Props as SelectItemProps, FProps as SelectItemFProps, SBProps as SelectItemSBProps } from "./Item";
export { SelectModalAligns } from "./Modal";
export type { FProps as SelectModalFProps, Props as SelectModalProps } from "./Modal";
export type { FProps as SelectOptionsFProps } from "./Options";
export type { FProps as SelectFProps, Props as SelectProps, SBProps as SelectSBProps } from "./Select";
export type { FProps as SelectCountriesFProps, Props as SelectCountriesProps } from "./Countries";

export default Select;
