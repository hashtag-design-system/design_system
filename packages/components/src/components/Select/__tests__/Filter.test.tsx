import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import Select from "../index";
import { selectCustomRender } from "../__helpers__/utils";
import { SelectTestChildren } from "./Select.test";

describe("<Select.Filter />", () => {
  // More tests of filter functionality in the "./Select.test.tsx" file
  test("default behaviour", async () => {
    selectCustomRender(<Select.Filter placeholder="Filter" floatingplaceholder={false} />);
    const input = screen.getByTestId("select-filter");

    expect(input).toBeVisible();
    expect(input).toHaveAttribute("class");
    expect(input.onchange).toBeDefined();
    expect(input.onkeydown).toBeDefined();
    expect(input).toHaveAttribute("placeholder", "Filter");
  });
  test("default behaviour inside <Select />", async () => {
    render(
      <Select defaultOpen>
        <SelectTestChildren />
      </Select>
    );
    const filterInput = screen.getByTestId("select-filter");

    await waitFor(() => {
      expect(filterInput).toBeVisible();
    });
    const testVal = "george";

    userEvent.type(filterInput, testVal);

    expect(filterInput).toHaveValue(testVal);
    await waitFor(() => {
      const items = screen.queryAllByTestId("select-item");
      expect(items).toHaveLength(1);
      expect(items[0]).toHaveTextContent("georgekrax");
    });
  });
  test("with filterById={true}", async () => {
    render(
      <Select defaultOpen>
        <SelectTestChildren filterById>
          <Select.Item id="georgekrax2" content="Amsterdam" />
        </SelectTestChildren>
      </Select>
    );
    const filterInput = screen.getByTestId("select-filter");

    await waitFor(() => {
      expect(filterInput).toBeVisible();
    });
    const testVal = "george";

    userEvent.type(filterInput, testVal);

    expect(filterInput).toHaveValue(testVal);
    await waitFor(() => {
      const items = screen.queryAllByTestId("select-item");
      expect(items).toHaveLength(2);
      expect(items[0]).toHaveTextContent("georgekrax");
      expect(items[1]).toHaveTextContent("Amsterdam");
    });
  });
  test("onChange basic functionality", () => {
    const onChange = jest.fn(e => e.target.value);
    render(
      <Select defaultOpen>
        <Select.Filter placeholder="Filter" onChange={e => onChange(e)} />
      </Select>
    );
    const input = screen.getByTestId("select-filter");
    const testVal = "george";

    userEvent.type(input, testVal);

    expect(input).toHaveValue(testVal);
    expect(onChange).toHaveBeenCalledTimes(testVal.length);
    expect(onChange.mock.results[testVal.length - 1].value).toBe(testVal);
  });
  describe("keyDown", () => {
    beforeEach(() => {
      render(
        <Select defaultOpen>
          <SelectTestChildren />
        </Select>
      );
    });
    test.each([false, true])("Tab & Tab + Escape", async bool => {
      const input = screen.getByTestId("select-filter");

      await waitFor(() => {
        expect(input).toBeVisible();
      });

      userEvent.click(input);
      expect(input).toHaveFocus();
      userEvent.tab();

      const firstItem = screen.getAllByTestId("select-item")[0];

      await waitFor(() => {
        expect(firstItem).toHaveFocus();
      });

      if (!bool) {
        userEvent.tab({ shift: true });
      } else {
        fireEvent.keyDown(firstItem, {
          key: "Escape",
          code: "Escape",
          keyCode: 27,
          charCode: 27,
        });
      }
      expect(screen.getByTestId("select-filter")).not.toHaveClass("focus");
    });
    test("Escape", async () => {
      const input = screen.getByTestId("select-filter");

      await waitFor(() => {
        expect(input).toBeVisible();
      });

      userEvent.type(input, "hey");
      fireEvent.keyDown(input, {
        key: "Escape",
        code: "Escape",
        keyCode: 27,
        charCode: 27,
      });
      // userEvent.type(input, specialChars.escape);
      await waitFor(() => {
        expect(screen.getByTestId("select-modal")).not.toBeVisible();
      });
      expect(screen.getByTestId("select")).not.toHaveAttribute("open");
      expect(input).not.toHaveFocus();
    });
    test("Enter", async () => {
      const input = screen.getByTestId("select-filter");

      await waitFor(() => {
        expect(input).toBeVisible();
      });

      userEvent.type(input, "hey");
      fireEvent.keyDown(input, {
        key: "Enter",
        code: "Enter",
        keyCode: 13,
        charCode: 13,
      });
      await waitFor(() => {
        expect(screen.getByTestId("select-modal")).not.toBeVisible();
      });
      expect(screen.getByTestId("select")).not.toHaveAttribute("open");
      await waitFor(() => {
        expect(screen.getByTestId("select-btn")).toHaveTextContent(screen.getAllByTestId("select-item")[0].textContent || "");
      });
    });
  });
});
