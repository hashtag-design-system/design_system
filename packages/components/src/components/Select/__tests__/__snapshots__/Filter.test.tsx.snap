// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Select /> with children with children <Select.Header /> 1`] = `
<div
  class="select__header"
  data-children="true"
  data-testid="select-header"
>
  <h6>
    Test header
  </h6>
</div>
`;
